File: backend/models/WorkoutSet.js
import mongoose from "mongoose";

const workoutSetSchema = new mongoose.Schema({
  exerciseId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "Exercise",
    required: true,
  },
  workoutName: {
    type: String,
    default: "Workout",
  },
  date: {
    type: Date,
    default: Date.now,
    required: true,
  },
  // 訓練組數記錄
  sets: [
    {
      weight: {
        type: Number,
        required: true,
      },
      reps: {
        type: Number,
        required: true,
      },
      isPersonalRecord: {
        type: Boolean,
        default: false,
      },
    },
  ],
  // 計算欄位
  calculated1RM: {
    type: Number,
  },
  totalVolume: {
    type: Number,
  },
});

// 前置保存鉤子自動計算 1RM 和 volume
//每當你 workoutSet.save() 時，它都會自動先跑這段程式碼，再去真的把資料存進 MongoDB。
workoutSetSchema.pre("save", function (next) {
  // 計算總訓練量
  this.totalVolume = this.sets.reduce(
    (sum, set) => sum + set.weight * set.reps,
    0
  );

  // 計算最大 1RM（使用 Brzycki 公式: weight * (36 / (37 - reps))）
  const oneRMs = this.sets.map((set) => set.weight * (36 / (37 - set.reps)));
  this.calculated1RM = Math.max(...oneRMs);

  next();
});

export default mongoose.model("WorkoutSet", workoutSetSchema);

File: mobile/lib/models/workout_set.dart
import 'dart:convert';

class WorkoutSet {
  final String id;
  final String exerciseId;
  final String workoutName;
  final DateTime date;
  final List<SetInfo> sets;
  final double? calculated1RM;
  final double? totalVolume;

  WorkoutSet({
    required this.id,
    required this.exerciseId,
    required this.workoutName,
    required this.date,
    required this.sets,
    this.calculated1RM,
    this.totalVolume,
  });

  factory WorkoutSet.fromJson(Map<String, dynamic> json) {
    return WorkoutSet(
      id: json['_id'],
      exerciseId: json['exerciseId'],
      workoutName: json['workoutName'] ?? '未命名訓練',
      date: DateTime.parse(json['date']),
      sets: (json['sets'] as List).map((e) => SetInfo.fromJson(e)).toList(),
      calculated1RM: json['calculated1RM']?.toDouble(),
      totalVolume: json['totalVolume']?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'exerciseId': exerciseId,
      'workoutName': workoutName,
      'date': date.toIso8601String(),
      'sets': sets.map((s) => s.toJson()).toList(),
      'calculated1RM': calculated1RM,
      'totalVolume': totalVolume,
    };
  }
}

class SetInfo {
  final double weight;
  final int reps;
  final bool isPersonalRecord;

  SetInfo({
    required this.weight,
    required this.reps,
    this.isPersonalRecord = false,
  });

  factory SetInfo.fromJson(Map<String, dynamic> json) {
    return SetInfo(
      weight: json['weight']?.toDouble() ?? 0.0,
      reps: json['reps'] ?? 0,
      isPersonalRecord: json['isPersonalRecord'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'weight': weight,
      'reps': reps,
      'isPersonalRecord': isPersonalRecord,
    };
  }
}
